### ISSUE TITLE
menu支持传入树结构数据

### ISSUE THREAD

**Issue by 102010cncger** [NG-ZORRO/ng-zorro-antd#364]
对angular4不是很熟悉,angularjs我可能就用$compile来实现了,
在这个版本如何根据下面的数据显示菜单呢?或者后续是否有支持
类似
``````````
<ul nz-menu [nzTheme]="'dark'" [nzMode]="isCollapsed?'vertical':'inline'" [menus]="menus">
<ul>
`````````````````````
````````````
export const MENUS: Menu[] = [
  {id: 'dashboard', name: 'Dashboard', url: '/dashboard', childs: []},
  {
    id: 'erros', name: 'Erro Page', url: '', childs:
    [
      {id: '404', name: '404', url: '/404', childs: []},
      {id: '500', name: '500', url: '/500', childs: [{id: 'dashboard', name: 'Dashboard', url: '/dashboard', childs: []},
        {
          id: 'erros1', name: 'Erro Page', url: '', childs:
          [
            {id: '4041', name: '404', url: '/404', childs: []},
            {id: '5001', name: '500', url: '/500', childs: [{
              id: 'erros1', name: 'Erro Page', url: '', childs:
                [
                  {id: '4041', name: '404', url: '/404', childs: []},
                  {id: '5001', name: '500', url: '/500', childs: [{
                    id: 'erros1', name: 'Erro Page', url: '', childs:
                      [
                        {id: '4041', name: '404', url: '/404', childs: []},
                        {id: '5001', name: '500', url: '/500', childs: []}
                      ]
                  }]}
                ]
            }]}
          ]
        }]}
    ]
  }
];
```````````````````

**Comment by vthinkxie** [comment]
use *ngFor please

**Comment by 102010cncger** [comment]
not effect perfect, all element Left margin is 24px. please help

i use like this
`````````````
<ul nz-menu [nzTheme]="'dark'" [nzMode]="isCollapsed?'vertical':'inline'">
      <li app-recursive-menu [menus]="menus"></li>
</ul>
````````````````````

`````````````````
import {Component, Input, OnInit} from '@angular/core';

@Component({
  selector: '[app-recursive-menu]',
  templateUrl: './recursive-menu.component.html',
  styleUrls: ['./recursive-menu.component.css']
})
export class RecursiveMenuComponent implements OnInit {

  constructor() {
  }

  ngOnInit() {
  }
  
  @Input() menus: any;
}

<ng-container *ngFor="let menu of menus">
  <li nz-menu-item *ngIf="!(menu.childs&&menu.childs.length>0)"><a routerLink="{{menu.url}}">{{menu.name}}</a></li>
  <li nz-submenu *ngIf="menu.childs&&menu.childs.length>0">
    <span title><i class="anticon anticon-team"></i><span class="nav-text">Team11111111111</span></span>
    <ul app-recursive-menu [menus]="menu.childs">
    </ul>
  </li>
</ng-container>
``````````````````

**Comment by vthinkxie** [comment]
Hi @102010cncger , I just checked the source code, subMenu can not be accessed when wrapped in  the other component, so the padding level is wrong with recursive component.
https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/src/components/menu/nz-submenu.component.ts#L152
You may need to handle the padding style yourself for the time being.
We will talk about it later, thanks for your feedback.

**Comment by noob9527** [comment]
@102010cncger @vthinkxie 

I just found a tricky way to do this,
by the way, I don't exactly know what does **ɵen** mean, 
why export internal component use these symbols?
Is there anyone can explain that to me?

```javascript
import { AfterViewInit, Component, Input, OnInit, QueryList, ViewChildren } from '@angular/core';
import { ɵen as NzSubMenuComponent } from 'ng-zorro-antd';

@Component({
  selector: 'app-menu-item',
  template: `
    <li nz-menu-item *ngIf="!menuItem.children">
      <a href="">
        {{menuItem.title}}
      </a>
    </li>
    <li nz-submenu *ngIf="menuItem.children">
      <span title>
        {{menuItem.title}}
      </span>
      <ul>
        <app-menu-item
          [menuItem]="item"
          *ngFor="let item of menuItem.children"
        >
        </app-menu-item>
      </ul>
    </li>
  `,
  styleUrls: ['./menu-item.component.css']
})
export class MenuItemComponent implements AfterViewInit {

  level = 1;
  @Input() menuItem: any;
  @ViewChildren(NzSubMenuComponent) nzSubMenus: QueryList<NzSubMenuComponent>;
  @ViewChildren(MenuItemComponent) subMenus: QueryList<MenuItemComponent>;

  ngAfterViewInit() {
    if (this.subMenus.length) {
      this.subMenus
        .filter(x => x !== this)
        .forEach(menu => {
          setTimeout(_ => {
            menu.level = this.level + 1;
            menu.syncNzSubMenusLevel();
          });
        });
    }
  }

  syncNzSubMenusLevel() {
    if (this.nzSubMenus.length) {
      this.nzSubMenus
        .forEach(menu => {
          menu.level = this.level;
        });
    }
  }

}

```

**Comment by vthinkxie** [comment]
@noob9527 Thanks a lot. ɵen is name after compile, it will be fixed next version.

**Comment by LinBoLen** [comment]
`ɵen`  the `ɵ` is used for private class export

**Comment by dyh333** [comment]
mark it

**Comment by noob9527** [comment]
follow by my previous comment
in the latest release(0.6.7), you can import **NzSubMenuComponent** directly from ''ng-zorro-antd'', so u should change this line

`import { ɵen as NzSubMenuComponent } from 'ng-zorro-antd';`
to
`import { NzSubMenuComponent } from 'ng-zorro-antd';`

It looks better, Although I still don't recommend since **NzSubMenuComponent** should be considered as "private" component.

**Comment by hatakawas** [comment]
has the recursive component been supported?

**Comment by vthinkxie** [comment]
try to fixed in version 0.7.0 but failed
https://github.com/NG-ZORRO/ng-zorro-antd/blob/5c0716107e0955b034a2d9ce75463ff3b0e9fdb6/components/menu/style/index.less#L372
the design of ant design can not insert any self define component into menu, it will break the style define.

**Comment by zhhb** [comment]
I has the same problem，thx for @noob9527  @vthinkxie

**Comment by candy4290** [comment]
i have the same problem.
<ng-container *ngFor="let menu of menus">
        <ng-container *ngTemplateOutlet="menuTpl; context: {$implicit: menu}"></ng-container>
      </ng-container>

      <ng-template #menuTpl let-menu>
        <li nz-menu-item *ngIf="!menu.items" [nzDisabled]="menu.disabled" [nzSelected]="menu.selected" routerLink="/route1" routerLinkActive="ant-menu-item-selected">
          <span title>
            <i [class]="menu.icon"></i>
            <span class="nav-text">{{menu.title}}</span>
          </span>
        </li>
        <li nz-submenu *ngIf="menu.items" [nzOpen]="menu.open" [nzDisabled]="menu.disabled">
          <span title>
            <i [class]="menu.icon"></i>
            <span class="nav-text">{{menu.title}}</span>
          </span>
          <ul>
            <ng-container *ngFor="let item of menu.items">
              <ng-container *ngTemplateOutlet="menuTpl; context: {$implicit: item}"></ng-container>
            </ng-container>
          </ul>
        </li>
      </ng-template>

-------------------------

this.menus = [(
      {
        title: '用户',
        icon: 'anticon anticon-user',
        open: true,
        selected: false,
        disabled: false,
        items: [
          {
            title: '用户1',
            icon: 'anticon anticon-user',
            selected: false,
            disabled: false,
            items: [
              {
                title: '用户11',
                icon: 'anticon anticon-user',
                selected: true,
                disabled: false
              },
              {
                title: '用户12',
                icon: 'anticon anticon-user',
                selected: false,
                disabled: false
              }
            ]
          },
          {
            title: '用户2',
            icon: 'anticon anticon-user',
            selected: false,
            disabled: false
          }
        ]
      },
      {
        title: 'Team',
        icon: 'anticon anticon-team',
        selected: false,
        disabled: false,
        items: [
          {
            title: 'Team1',
            icon: 'anticon anticon-team',
            selected: false,
            disabled: false,
          },
          {
            title: 'Team2',
            icon: 'anticon anticon-team',
            selected: false,
            disabled: false,
          }
        ]
      },
      {
        title: 'test',
        icon: 'anticon anticon-team',
        selected: false,
        disabled: false
      }
    )];

**Comment by dyh333** [comment]
does it fixed in the lastest version(v7.0.0)?

**Comment by vthinkxie** [comment]
@dyh333 plz read the latest doc
